### PR TITLE
Initial implementation of autograd

### LINKED ISSUE #2: Automatic Differentiation
This would be cool: http://en.wikipedia.org/wiki/Automatic_differentiation
Theano uses this.

This should be a secondary objective. Initially (and for performance reasons) it will make sense to provide the derivatives of functions, however this feature will provide extensibility for later unknown model types.

### PR DESCRIPTION
**Pull request by pavanky**
What is done so far:

- A proof of concept implementation for automatic differentiation using `autograd::Variable`, `autograd::backward`. 
- This currently implements only a few basic operations (only +, * for now).
- Ability to perform first order derivatives. Higher order derivatives to come later.

**Variable**
- This can be constructed in two ways
  - Using an `af::array` from the user
  - An operator returning a Variable. The operator constructs the Variable using a set of input Variables, the output array and a grad function.
- When `var.backward(grad_var)` is invoked, it builds a DAG as vector starting with the current variable and propagates gradients down the graph to all the Variables in the graph using the grad function specified at each variable.
- Calculating Gradients for a variable (and its subgraph) can be disabled by ivoking `var.setCalcGrad(false)` 

**Functions**
- Each function takes in `Variable` parameters and return `Variable` as a parameter.
- Each function performs the operation immediately on the data.
- Each function returns a `Variable` constructed using arguments as parameters:
  - `af::array`: The result calculated earlier
  - `vector<Variable>`: containing the inputs to the function
  -  `BackwardFunction_t`: A function pointer to the backward pass. Usually implemented as a lambda function.

Example function:
 ```c++
        Variable operator +(const Variable lhs, const Variable rhs)
        {
            auto result = lhs.getData() + rhs.getData();
            auto backward = [](std::vector<Variable> inputs, Variable grad_output) {
                inputs[0].addGrad(grad_output);
                inputs[1].addGrad(grad_output);
            };
            return Variable(result, {lhs, rhs}, backward);
        }
```
-------------------
**Example**:

A simple example showcasing how this can be done currently

```c++
void test()
{
    using af::autograd::Variable;
    auto x = Variable(af::randu(5), true);
    af_print(x.array());
    auto y = Variable(af::randu(5), true);
    af_print(y.array());
    auto z = x * x + x * y + y * y;
    auto dz = Variable(af::constant(1.0, 5), false);
    z.backward(dz);
    auto dx = x.grad();
    auto dy = y.grad();
    af_print(dx.array() - 2 * x.array() - y.array());
    af_print(dy.array() - 2 * y.array() - x.array());
}
```

-----------------
**TODO:** for this PR

- [x] Add all math operations: +, -, *, /, sin, cos, exp, tanh
- [x] Add array operations: tile, sum, transpose
- [x] Add operations required for Dense layers: matmul
- [x] Reimplement existing layers using autograd
- [x] Option to enable or disable building sub graphs
- [x] Option to enable or disable retaining graphs for gradients
- [x] Make sure perceptron example is working.
- [x] Add train and evaluation mode for modules
-----------------------
- Fixes #2

### CODE REVIEW

**Comment by pavanky** [comment]
@botev @jramapuram @itsnarsi This has been a long time coming, but I'd appreciate if you guys had any feedback as well.

**Comment by pavanky** [comment]
CC @arrayfire/core-devel

**Comment by pavanky** [comment]
@Reithan too

**Comment by jramapuram** [comment]
Awesome work @pavanky . Will take a look in more detail when I get to a terminal. Quick question: can you take second derivatives with your implementation?

**Comment by pavanky** [comment]
@jramapuram Not yet, I wanted to get the first order working first :)

**Comment by pavanky** [comment]
@jramapuram went ahead and changed the gradients to be Variables too. This should make it easy to perform higher order derivatives.

**Comment by itsnarsi** [comment]
@pavanky just tested it on my laptop and it looks pretty neat. Unlike python, I did not see any initial delay. This might be because of no JIT I guess.
When will this be merged to this repo?

File: examples/FFNet.cpp [hunk 1]
```diff
@@ -7,14 +7,15 @@
  * http://arrayfire.com/licenses/BSD-3-Clause
  ********************************************************/
 
-#include <afml/nn.h>
+#include <af/nn.h>
 
 using namespace af;
-using namespace afml;
-using namespace afml::nn;
+using namespace af;
```
Comment (FloopCZ) [inline examples/FFNet.cpp]: Duplicated line

File: include/af/autograd/Variable.hpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                std::vector<Variable> getInputs()
+                {
+                    return m_inputs;
+                }
+
+                void evalGrad()
+                {
+                    if (m_grads.size() == 1) return;
+                    Variable grad = m_grads[0];
+                    for (int i = 1; i < (int)m_grads.size(); i++) {
```
Comment (FloopCZ) [inline include/af/autograd/Variable.hpp]: I would prefer unsigned iterable to avoid clang's `-Wconversion` signedness warnings when indexing to `std::vector`.

**Comment by pavanky** [comment]
@itsnarsi This is still very nascent. I want to incorporate some of the stuff mentioned here to make it more efficient: 
http://pytorch.org/docs/master/notes/autograd.html#excluding-subgraphs

**Review by FloopCZ — COMMENTED** [review]
Hmm, nice job!

File: examples/FFNet.cpp (hunk 1 above)
Comment (pavanky) [reply examples/FFNet.cpp]: Do you have a tool for detecting this or a really good eye :D

File: include/af/autograd/Variable.hpp (hunk 2 above)
Comment (pavanky) [reply include/af/autograd/Variable.hpp]: Will do thanks.

File: examples/FFNet.cpp (hunk 1 above)
Comment (FloopCZ) [reply examples/FFNet.cpp]: A tool would be great. Unfortunately, I'm just an irritating nitpicker. :innocent:

**Comment by pavanky** [comment]
Decreased the scope of the PR to get a minimum viable thing going. The additional functions and operators can be added once this PR gets merged.

**Comment by pavanky** [comment]
@jramapuram I think enabling the support for higher order derivatives by default will increase the memory being used. I am going to enable a flag to enable it during the backward pass. By default only the values will be stored.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,26 +1,36 @@
...
-)
+cmake_minimum_required(VERSION 3.5.1)
+
+project(ArrayFireML
+  VERSION 0.1.0
+  LANGUAGES C CXX)
+
+find_package(ArrayFire REQUIRED)
+
+add_library(afml SHARED "")
```
Comment (umar456) [inline CMakeLists.txt:9]: If you don't add SHARED then you can control the type of library you make by BUILD_SHARED_LIBS variable

File: src/autograd/Functions.cpp [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+#include <af/autograd/Variable.hpp>
+#include <af/autograd/Functions.hpp>
+
+namespace af {
+    namespace autograd {
+
+        Variable operator +(const Variable &lhs, const Variable &rhs)
+        {
+            auto result = lhs.array() + rhs.array();
+            auto grad_func = [](std::vector<Variable> &inputs, const Variable &grad_output) {
```
Comment (umar456) [inline src/autograd/Functions.cpp:19]: Don't we usually have outputs then inputs?

It looks like you know the # of inputs for each function. I would use something like std::array<Variable, N> for something like that

**Review by umar456 — COMMENTED** [review]
Minor preliminary comments. Everything looks great. We can refactor it later as long as we have a clean user-facing API.

File: src/autograd/Functions.cpp (hunk 4 above)
Comment (pavanky) [reply src/autograd/Functions.cpp:19]: Both of these are inputs. `grad_output` is an input coming from a different place.

Comment (pavanky) [reply src/autograd/Functions.cpp:19]: And using `std::array` is not an option. All functions need to share the same signature so they can be stored as `GradFunc_t` inside Variable.

File: examples/perceptron.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            // TODO: Use loss function
+            af::array diff = out_j - result.array();
+
+            // Backward propagation
+            auto d_result = Variable(diff, false);
+            result.backward(d_result);
+
+            // Update parameters
+            // TODO: Should use optimizer
+            for (auto param : perceptron.parameters()) {
```
Comment (umar456) [inline examples/perceptron.cpp]: auto& ?

File: examples/perceptron.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
+/*******************************************************
+ * Copyright (c) 2015, ArrayFire
```
Comment (umar456) [inline examples/perceptron.cpp]: 2017

File: include/af/autograd/Variable.hpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                       bool calc_grad);
+
+                bool m_calc_grad;
+                af::array m_data;
+                std::vector<Variable> m_inputs;
+                std::vector<Variable> m_grads;
+                GradFunc_t m_grad_func;
+            };
+
+            public:
```
Comment (umar456) [inline include/af/autograd/Variable.hpp]: Needs to be aligned with other access qualifiers.

File: src/nn/Modules/Module.cpp [hunk 8]
```diff
@@ -0,0 +1,61 @@
+/*******************************************************
+ * Copyright (c) 2015, ArrayFire
```
Comment (umar456) [inline src/nn/Modules/Module.cpp]: 2017

File: src/nn/Modules/Module.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        void Module::train()
+        {
+            for (auto parameter : m_parameters) {
+                parameter.setCalcGrad(true);
+            }
+        }
+
+        void Module::eval()
+        {
+            for (auto parameter : m_parameters) {
```
Comment (umar456) [inline src/nn/Modules/Module.cpp]: auto&?

File: include/af/autograd/Variable.hpp [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+            void calcGradInputs(bool retain_grad_graph = false);
+
+            void backward(const Variable &grad, bool retain_grad_graph = false);
+
+
+        private:
+            void evalGrad(bool retain_grad_graph = false);
+
+            std::vector<Variable> getInputs() const;
```
Comment (umar456) [inline include/af/autograd/Variable.hpp]: Does this need to return by value?

**Review by umar456 — APPROVED** [review]
A couple of minor issues. This is looking great!
